### DISCUSSION TITLE
Issues with Kubernetes Mock Server since 6.4.0

### DISCUSSION THREAD

**Discussion by scholzj** [fabric8io/kubernetes-client#4924] [Q&A]
In Strimzi, we use for some of our tests the Kubernetes Mock Server and it worked reasonably fine for us until 6.3.1. But from 6.4.0, we seem to have two issues in many places:
* The tests are approximately 4-5 times slower compared to 6.3.1
* When calling `replaceStatus()` or `patchStatus()` on one of our custom resources, it seems there is a race condition where for some time, the mock server returns `null` / `404` to a `get()` call on a given resource which is done in parallel to the status update call.

This seems to happen with both 6.4.0 and 6.4.1 but also with the latest 6.5-SNAPSHOT builds.

Is this a known issue? Or anyone else experienced something similar? If not, I will try to put together some reproducer.

**Comment by manusa** [comment]
> The tests are approximately 4-5 times slower compared to 6.3.1

This was also reported in #4786. We did have some changes recently that involved additional serialization-deserialization overhead (#4801), and also added resource version validation (for optimistic locking) in #4740. However, none of these have been released yet. If the difference in performance is between 6.4.x and 6.5-SNAPSHOT, then these are good candidates.

If the performance degradation also appears in 6.4 it can only be related to the changes in OkHttp client-side.

> it seems there is a race condition

For the new race condition, if it appears in 6.5 then can be related to #4740.

In any case, having a few more hints would help in resolving these issues.

**Reply by scholzj** [reply]
> If the performance degradation also appears in 6.4 it can only be related to the changes in OkHttp client-side.

I see the performance degradation already in 6.4. But we use the Java HTTP client now.

I will try to put together something less dependent on Strimzi so that it is easier to share.

**Reply by shawkins** [reply]
> I see the performance degradation already in 6.4. But we use the Java HTTP client now.

Drilling down on this a bit more using the client kubernetes-tests I see that the big difference is PodExecTest - that's increasing the test runtime by 30 seconds.  The issue is the change to more broadly apply a readyWaitTimeout to all pod operations (exec, logs, copy, etc.) and not just logging.  Against the mock server this is problematic because the Pod may not be setup with a ready status, so more than likely it's just waiting for the timeout to expire.  Can you confirm if your issue is around pod operations?

On #4786 that seems to be about a test slow down in general - it's not something that we see clearly.

With a minor fix and disabling the PodTest, PodExecTest, and BuildConfigTest (which cannot be run against all of the clients), running our kubernetes-tests against master with jdk 17 locally gave me:

okhttp: 69 seconds
jetty: 76 seconds
jdk: 70 seconds
vertx: 75 seconds

With similar execution times against 6.3.1.  So excluding PodExecTest there doesn't seem to be much difference across clients or versions in terms of general performance.  That issue was referencing differences in the milliseconds, which could be attributable to running with debug logging turned on.

**Reply by scholzj** [reply]
We do not do any exec, copy, or logs in our tests. But in many of our tests using the mock server, there are several asynchronous components interacting with each other through the mock server. So I think it is possible that some of that amplifies the slowdown. But the tips are good as I know better what to look for.

**Reply by shawkins** [reply]
Do you have a pointer to what strimzi module and tags / branches I could use for comparison?  If this can be identified, a fix could go into 6.5.

**Reply by scholzj** [reply]
Sorry, I was busy with something else. I created this as a reproducer:

```java
package io.strimzi.operator.cluster.operator.assembly;

import io.fabric8.kubernetes.api.model.ContainerBuilder;
import io.fabric8.kubernetes.api.model.Pod;
import io.fabric8.kubernetes.api.model.PodBuilder;
import io.fabric8.kubernetes.client.KubernetesClient;
import io.fabric8.kubernetes.client.server.mock.EnableKubernetesMockClient;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.junit.jupiter.api.Test;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CountDownLatch;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.notNullValue;

@EnableKubernetesMockClient(crud = true)
public class Fabric8Test {
    private static final Logger LOGGER = LogManager.getLogger(Fabric8Test.class);

    private static final String NAME = "testpod";
    private static final Pod POD = new PodBuilder()
            .withNewMetadata()
                .withName(NAME)
            .endMetadata()
            .withNewSpec()
                .withContainers(new ContainerBuilder()
                        .withName("busybox")
                        .withImage("quay.io/scholzj/busybox:latest") // Quay.io is used to avoid Docker Hub limits
                        .withCommand("sleep", "3600")
                        .withImagePullPolicy("IfNotPresent")
                        .build())
                .withRestartPolicy("Always")
                .withTerminationGracePeriodSeconds(0L)
            .endSpec()
            .build();

    // Injected by Fabric8 Mock Kubernetes Server
    @SuppressWarnings("unused")
    private KubernetesClient client;

    @Test
    public void testMockServerRaceCondition()   {
        CountDownLatch latch = new CountDownLatch(1);

        // First we create
        client.pods().resource(POD).create();

        // Check that it exists
        assertThat(client.pods().withName(NAME).get(), is(notNullValue()));

        // Update pod status of the pod 100 times in a loop
        CompletableFuture.runAsync(() -> {
            for (int i = 0; i < 100; i++) {
                LOGGER.info("Updating Pod status: loop {}", i);
                long startTime = System.nanoTime();

                client.pods()
                        .resource(new PodBuilder(POD)
                                .editOrNewStatus()
                                .withMessage("Status revision " + i)
                                .endStatus()
                                .build())
                        .replaceStatus();

                long elapsedTime = System.nanoTime() - startTime;
                LOGGER.info("Pod status updated in {} ms", elapsedTime/1000000);
            }

            latch.countDown();
        });

        // Get the pod in a loop to see if it exists
        while (true)    {
            LOGGER.info("Asserting pod");
            assertThat(client.pods().withName(NAME).get(), is(notNullValue()));

            if (latch.getCount() == 0)  {
                break;
            }
        }

        LOGGER.info("The test is complete");
    }
}
```

(If you want, I can provide it as a separate project on GitHub - but it now doesn't depend on Strimzi so you should be basically just able to copy it and run it.)

It basically creates a pod, in a separate thread and updates its status in a loop while in the test it periodically gets it and asserts that it is not null.

I observed the following:
* You are right, it fails with 6.3.1 as well. But the `replaceStatus` request seems to be faster there (usually 2 - 7 ms) so it runs into the race only sometimes for example only after 50th loop.
* With 6.4.1 and with 6.5-SNAPSHOT, the `replaceStatus` call seems to take about 15 - 20 ms. That makes the test usually fail already in the first or second loop.

**Reply by scholzj** [reply]
This is an example of an actual Strimzi test where it is failing with 6.4 and passing with 6.3: https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorCustomCertMockTest.java => but as the reproducer suggest that is likely just because of the timing which is different. This test runs the Strimzi controllers + our own _mock controllers_ which emulate the Pod or Service readiness in the MockKube. So they are more complicated compared to the reproducer.

**Reply by shawkins** [reply]
The fix in #4945 does address the null issue.  I don't see as big of a performance difference locally, but enough of a difference between 6.3.1 and 6.4.1 that something should be attributable.

**Reply by shawkins** [reply]
The biggest difference locally was a change to move the Serialization logic handling multiple docs in a single stream to a more common location.  The doc validate method probably because of the negative lookahead, is quite expensive even over relatively short inputs.

**Reply by scholzj** [reply]
Thanks a lot for looking into it.

**Reply by scholzj** [reply]
I tried it with 6.5.0 and all the issues I reported here are fixed. Also the speed seems to be (at least roughly) the same as before with 6.3.1. Thanks a lot for that 😍. 

Unfortunately I run into another issue with leader election and the Mock Server - but that seems to completely unrelated to this. I opened https://github.com/fabric8io/kubernetes-client/issues/4965 for it.

**Comment by shawkins** [comment]
> When calling replaceStatus() or patchStatus() on one of our custom resources, it seems there is a race condition where for some time, the mock server returns null / 404 to a get() call on a given resource which is done in parallel to the status update call.

A possible root issue is that there is inconsistent synchronization around the map of resources held by the mock server.  The KubernetesCrudDispatcher.processEvent method will temporarily remove the resource, then add it back if it's not a delete operation so there is a possible race condition for patch and put operations.  It looks like create, delete, get, and watches have appropriate locking when being processed.  That doesn't seem like a new problem though - the logic has been this way at least since 6.0.

**Reply by scholzj** [reply]
Well, I never saw it before with 6.3.1 or older. But as I said, things seem to be also much slower now with 6.4.0+. So it could be that this just made it much more likely for it to happen and it was just a coincidence it seemed to work fine before.

**Reply by shawkins** [reply]
That's possible - we noticed with other testing that the okhttp client seems to some implicit flow control that would potentially avoid some concurrency issues.
